File: stock_charts/stock_charts.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <memory>

using std::vector;
using std::cin;
using std::cout;

class StockCharts {
 public:
  void Solve() {
    vector<vector<int>> stock_data = ReadData();
    int result = MinCharts(stock_data);
    WriteResponse(result);
  }

 private:
  vector<vector<int>> ReadData() {
    int num_stocks, num_points;
    cin >> num_stocks >> num_points;
    vector<vector<int>> stock_data(num_stocks, vector<int>(num_points));
    for (int i = 0; i < num_stocks; ++i)
      for (int j = 0; j < num_points; ++j) {
        cin >> stock_data[i][j];
      }
    return stock_data;
  }

  void WriteResponse(int result) {
    cout << result << "\n";
  }

  vector<vector<bool>> converter(const vector<vector<int>>& stock_data,int n,int k){
    vector<vector<bool>> adj_matrix(n,vector<bool>(n,true));
    for(int i=0;i<n;++i){
      for(int j=0;j<n;++j){
        for(int t=0;t<k;++t){
          if(stock_data[i][t]>=stock_data[j][t]){
            adj_matrix[i][j] = false;
            break;
          }
        }
      }
    }
    return adj_matrix;
  }

  bool dfs(const vector<vector<bool>>& adj_matrix,vector<int>& matching, vector<bool>& seen,int u,int n){
    for(int v=0;v<n;++v){
      if(adj_matrix[u][v] && !seen[v]){
        seen[v] = true;
        if(matching[v]<0 || dfs(adj_matrix,matching,seen,matching[v],n)){
          matching[v] = u;
          return true;
        }
      }
    }
    return false;
  }

  int MinCharts(const vector<vector<int>>& stock_data) {
    // Replace this incorrect greedy algorithm with an
    // algorithm that correctly finds the minimum number
    // of charts on which we can put all the stock data
    // without intersections of graphs on one chart.

    int n = stock_data.size();
    int k = stock_data[0].size();
    // Vector of charts; each chart is a vector of indices of individual stocks.
    vector<vector<bool>> adj_matrix = converter(stock_data,n,k);
    vector<int> matching(n,-1);
    int ans=0;
    for(int u=0;u<n;++u){
      vector<bool> seen(n,false);
      if(dfs(adj_matrix,matching,seen,u,n))
        ans++;
    }
    
    return n-ans;
  }
};

int main() {
  std::ios_base::sync_with_stdio(false);
  StockCharts stock_charts;
  stock_charts.Solve();
  return 0;
}

File: stock_charts/stock_charts.py
# python3
class StockCharts:
    def read_data(self):
        n, k = map(int, input().split())
        stock_data = [list(map(int, input().split())) for i in range(n)]
        return stock_data

    def write_response(self, result):
        print(result)

    def converter(self,stock_data,n,k):
        adj_matrix = [[0]*n for _ in range(n)]
        for i in range(n):
            for j in range(n):
                Isedge = True
                for t in range(k):
                    if(stock_data[i][t]>=stock_data[j][t]):
                        Isedge = False
                        break
                if Isedge:
                    adj_matrix[i][j] = 1
        return adj_matrix

    def dfs(self, adj_matrix,matching,seen,u,n):
        for v in range(n):
            if adj_matrix[u][v]==1 and seen[v]==False:
                seen[v] = True
                if(matching[v]<0 or self.dfs(adj_matrix,matching,seen,matching[v],n)):
                    matching[v] = u
                    return True
        return False


    def min_charts(self, stock_data):
        # Replace this incorrect greedy algorithm with an
        # algorithm that correctly finds the minimum number
        # of charts on which we can put all the stock data
        # without intersections of graphs on one chart.
        n = len(stock_data)
        k = len(stock_data[0])
        adj_matrix = self.converter(stock_data,n,k)
        matching = [-1]*n
        ans = 0
        for u in range(n):
            seen = [False]*n 
            if(self.dfs(adj_matrix,matching,seen,u,n)):
                ans += 1
        return n-ans

    def solve(self):
        stock_data = self.read_data()
        result = self.min_charts(stock_data)
        self.write_response(result)

if __name__ == '__main__':
    stock_charts = StockCharts()
    stock_charts.solve()
